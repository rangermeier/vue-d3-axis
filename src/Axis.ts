import { PropType } from 'vue'
import type { Scaler, AxisStyle } from './types'

function translateX<T>(scale0: Scaler<T>, scale1: Scaler<T>, d: T): string {
  const x = scale0(d)
  return `translate(${isFinite(x) ? x : scale1(d)},0)`
}

function translateY<T>(scale0: Scaler<T>, scale1: Scaler<T>, d: T): string {
  const y = scale0(d)
  return `translate(0,${isFinite(y) ? y : scale1(d)})`
}

export const TOP = 'TOP'
export const RIGHT = 'RIGHT'
export const BOTTOM = 'BOTTOM'
export const LEFT = 'LEFT'

const defaultAxisStyle: AxisStyle = {
  orient: BOTTOM,
  tickSizeInner: 6,
  tickSizeOuter: 6,
  tickPadding: 3,
  strokeWidth: 1,
  strokeColor: 'currentColor',
  tickFont: 'sans-serif',
  tickFontSize: 10
}

export default {
  render: function (createElement) {
    const axisStyle = { ...defaultAxisStyle, ...this.styles }
    const {
      orient,
      tickSizeInner,
      tickPadding,
      tickSizeOuter,
      strokeWidth,
      strokeColor,
      tickFont,
      tickFontSize
    } = axisStyle

    const transform = orient === TOP || orient === BOTTOM
      ? translateX
      : translateY
    const tickTransformer = (d): string => transform(this.position, this.position, d)

    const k = orient === TOP || orient === LEFT ? -1 : 1
    const isRight = orient === RIGHT
    const isLeft = orient === LEFT
    const isTop = orient === TOP
    const isBottom = orient === BOTTOM
    const isHorizontal = isRight || isLeft
    const x = isHorizontal ? 'x' : 'y'
    const y = isHorizontal ? 'y' : 'x'

    const halfWidth = strokeWidth / 2
    const range0 = this.range[0] + halfWidth
    const range1 = this.range[this.range.length - 1] + halfWidth

    const spacing = Math.max(tickSizeInner, 0) + tickPadding

    return createElement(
      'g',
      {
        attrs: {
          fill: 'none',
          fontSize: tickFontSize,
          fontFamily: tickFont,
          textAnchor: isRight ? 'start' : isLeft ? 'end' : 'middle',
          strokeWidth: strokeWidth
        }
      },
      [
        createElement('path', {
          attrs: {
            stroke: strokeColor,
            d: isHorizontal
              ? `M${k * tickSizeOuter},${range0}H${halfWidth}V${range1}H${k * tickSizeOuter}`
              : `M${range0},${k * tickSizeOuter}V${halfWidth}H${range1}V${k * tickSizeOuter}`
          }
        }),
        this.values.map((v, idx) => {
          const lineProps = { stroke: strokeColor }
          lineProps[`${x}2`] = k * tickSizeInner
          lineProps[`${y}1`] = halfWidth
          lineProps[`${y}2`] = halfWidth

          const textProps = {
            fill: strokeColor,
            dy: isTop ? '0em' : isBottom ? '0.71em' : '0.32em'
          }
          textProps[`${x}`] = k * spacing
          textProps[`${y}`] = halfWidth
          return createElement('g', {
            attrs: {
              key: `tick-${idx}`,
              opacity: 1,
              transform: tickTransformer(v)
            },
          }, [
            createElement('line', {
              attrs: lineProps
            }),
            createElement('text', {
              attrs: textProps
            }, this.format(v)),
          ])
        })
      ]
    )
  },
  props: {
    styles: {
      type: Object,
      required: false
    },
    range: {
      type: Array as PropType<number[]>,
      required: false
    },
    values: {
      type: Array as PropType<number[]>,
      required: false
    },
    position: {
      type: Function as PropType<Scaler<number>>,
      required: false
    },
    format: {
      type: Function as PropType<(d: number) => string>,
      required: false
    }
  }
}
