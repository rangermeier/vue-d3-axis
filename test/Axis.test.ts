import { shallowMount } from '@vue/test-utils'
import Axis, {TOP, RIGHT, BOTTOM, LEFT} from '@/Axis'


test('Renders bottom-oriented', () => {
  const wrapper = shallowMount(Axis, {
    propsData: {
      styles: {orient: BOTTOM},
      range: [0, 1000],
      values: [0, 250, 500],
      position: d => d * 2,
      format: d => `${d}`
    }
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders top-oriented', () => {
  const wrapper = shallowMount(Axis, {
    propsData: {
      styles: {orient: TOP},
      range: [0, 1000],
      values: [0, 250, 500],
      position: d => d * 2,
      format: d => `${d}`
    }
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders left-oriented', () => {
  const wrapper = shallowMount(Axis, {
    propsData: {
      styles: {orient: LEFT},
      range: [0, 1000],
      values: [0, 250, 500],
      position: d => d * 2,
      format: d => `${d}`
    }
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders right-oriented', () => {
  const wrapper = shallowMount(Axis, {
    propsData: {
      styles: {orient: RIGHT},
      range: [0, 1000],
      values: [0, 250, 500],
      position: d => d * 2,
      format: d => `${d}`
    }
  })
  expect(wrapper.element).toMatchSnapshot()
})
