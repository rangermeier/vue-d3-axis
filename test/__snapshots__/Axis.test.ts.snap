// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders bottom-oriented 1`] = `
<g
  fill="none"
  fontFamily="sans-serif"
  fontSize="10"
  strokeWidth="1"
  textAnchor="middle"
>
  <path
    d="M0.5,6V0.5H1000.5V6"
    stroke="currentColor"
  />
  <g
    key="tick-0"
    opacity="1"
    transform="translate(0,0)"
  >
    <line
      stroke="currentColor"
      x1="0.5"
      x2="0.5"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="currentColor"
      x="0.5"
      y="9"
    >
      0
    </text>
  </g>
  <g
    key="tick-1"
    opacity="1"
    transform="translate(500,0)"
  >
    <line
      stroke="currentColor"
      x1="0.5"
      x2="0.5"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="currentColor"
      x="0.5"
      y="9"
    >
      250
    </text>
  </g>
  <g
    key="tick-2"
    opacity="1"
    transform="translate(1000,0)"
  >
    <line
      stroke="currentColor"
      x1="0.5"
      x2="0.5"
      y2="6"
    />
    <text
      dy="0.71em"
      fill="currentColor"
      x="0.5"
      y="9"
    >
      500
    </text>
  </g>
</g>
`;

exports[`Renders left-oriented 1`] = `
<g
  fill="none"
  fontFamily="sans-serif"
  fontSize="10"
  strokeWidth="1"
  textAnchor="end"
>
  <path
    d="M-6,0.5H0.5V1000.5H-6"
    stroke="currentColor"
  />
  <g
    key="tick-0"
    opacity="1"
    transform="translate(0,0)"
  >
    <line
      stroke="currentColor"
      x2="-6"
      y1="0.5"
      y2="0.5"
    />
    <text
      dy="0.32em"
      fill="currentColor"
      x="-9"
      y="0.5"
    >
      0
    </text>
  </g>
  <g
    key="tick-1"
    opacity="1"
    transform="translate(0,500)"
  >
    <line
      stroke="currentColor"
      x2="-6"
      y1="0.5"
      y2="0.5"
    />
    <text
      dy="0.32em"
      fill="currentColor"
      x="-9"
      y="0.5"
    >
      250
    </text>
  </g>
  <g
    key="tick-2"
    opacity="1"
    transform="translate(0,1000)"
  >
    <line
      stroke="currentColor"
      x2="-6"
      y1="0.5"
      y2="0.5"
    />
    <text
      dy="0.32em"
      fill="currentColor"
      x="-9"
      y="0.5"
    >
      500
    </text>
  </g>
</g>
`;

exports[`Renders right-oriented 1`] = `
<g
  fill="none"
  fontFamily="sans-serif"
  fontSize="10"
  strokeWidth="1"
  textAnchor="start"
>
  <path
    d="M6,0.5H0.5V1000.5H6"
    stroke="currentColor"
  />
  <g
    key="tick-0"
    opacity="1"
    transform="translate(0,0)"
  >
    <line
      stroke="currentColor"
      x2="6"
      y1="0.5"
      y2="0.5"
    />
    <text
      dy="0.32em"
      fill="currentColor"
      x="9"
      y="0.5"
    >
      0
    </text>
  </g>
  <g
    key="tick-1"
    opacity="1"
    transform="translate(0,500)"
  >
    <line
      stroke="currentColor"
      x2="6"
      y1="0.5"
      y2="0.5"
    />
    <text
      dy="0.32em"
      fill="currentColor"
      x="9"
      y="0.5"
    >
      250
    </text>
  </g>
  <g
    key="tick-2"
    opacity="1"
    transform="translate(0,1000)"
  >
    <line
      stroke="currentColor"
      x2="6"
      y1="0.5"
      y2="0.5"
    />
    <text
      dy="0.32em"
      fill="currentColor"
      x="9"
      y="0.5"
    >
      500
    </text>
  </g>
</g>
`;

exports[`Renders top-oriented 1`] = `
<g
  fill="none"
  fontFamily="sans-serif"
  fontSize="10"
  strokeWidth="1"
  textAnchor="middle"
>
  <path
    d="M0.5,-6V0.5H1000.5V-6"
    stroke="currentColor"
  />
  <g
    key="tick-0"
    opacity="1"
    transform="translate(0,0)"
  >
    <line
      stroke="currentColor"
      x1="0.5"
      x2="0.5"
      y2="-6"
    />
    <text
      dy="0em"
      fill="currentColor"
      x="0.5"
      y="-9"
    >
      0
    </text>
  </g>
  <g
    key="tick-1"
    opacity="1"
    transform="translate(500,0)"
  >
    <line
      stroke="currentColor"
      x1="0.5"
      x2="0.5"
      y2="-6"
    />
    <text
      dy="0em"
      fill="currentColor"
      x="0.5"
      y="-9"
    >
      250
    </text>
  </g>
  <g
    key="tick-2"
    opacity="1"
    transform="translate(1000,0)"
  >
    <line
      stroke="currentColor"
      x1="0.5"
      x2="0.5"
      y2="-6"
    />
    <text
      dy="0em"
      fill="currentColor"
      x="0.5"
      y="-9"
    >
      500
    </text>
  </g>
</g>
`;
